src/processing/threshold.rs

```diff
@@ -0,0 +1,280 @@
+
+    /// Run the Otsu threshold detection algorithm with the 
+    /// given parameters. Due to Otsu being specified as working 
+    /// on greyscale images all current implementations
+    /// assume a single channel image returning an error otherwise.
+    fn threshold_otsu(&self) -> Result<Self::Output, Error>;
+}
+
+/// Runs the Mean Thresholding algorithm on a type T
+pub trait ThresholdMeanExt<T> {
+    /// Output type, this is different as Otsu outputs a binary image
+    type Output;
+
+    /// Run the Otsu threshold detection algorithm with the 
+    /// given parameters. Due to Otsu being specified as working 
+    /// on greyscale images all current implementations
+    /// assume a single channel image returning an error otherwise.
+    fn threshold_mean(&self) -> Result<Self::Output, Error>;
+}
+
+impl<T, C> ThresholdOtsuExt<T> for Image<T, C>
+where
+    Image<T, C>: Clone,
+    T: Copy + Clone + Ord + Num + NumAssignOps + ToPrimitive + FromPrimitive + PixelBound,
+    C: ColourModel,
+{
+    type Output = Image<bool, C>;
+
+    fn threshold_otsu(&self) -> Result<Self::Output, Error> {
+        let data = self.data.threshold_otsu()?;
+        Ok(Self::Output {
+            data,
+            model: PhantomData,
+        })
+    }
+}
+
+impl<T> ThresholdOtsuExt<T> for Array3<T>
+where
+    T: Copy + Clone + Ord + Num + NumAssignOps + ToPrimitive + FromPrimitive 
+{
+    type Output = Array3<bool>;
+
+    fn threshold_otsu(&self) -> Result<Self::Output, Error> {
+        if self.shape()[2] > 1 {
+            Err(Error::ChannelDimensionMismatch)
+        } else {
+            let value = calculate_threshold_otsu(&self)?;
+            let mask = apply_threshold(self, value);
+            Ok(mask)
+        }
+    }
+}
+
+///
+/// Calculates Otsu's threshold 
+/// Works per channel, but currently 
+/// assumes grayscale, i.e. single channel; otherwise need to 
+/// output all 3 values, or something!). 
+/// Todo: Add optioal nbins
+///
+fn calculate_threshold_otsu<T>(mat: &Array3<T>) -> Result<f64, Error>
+where
+    T: Copy + Clone + Ord + Num + NumAssignOps + ToPrimitive + FromPrimitive 
+{
+    let mut threshold = 0.0;
+    let n_bins = 255;
+    for c in mat.axis_iter(Axis(2)) {
+        let scale_factor = (n_bins) as f64
+            /(c.max().unwrap().to_f64().unwrap()); 
+        let edges_vec: Vec<u8> = (0..n_bins).collect();
+        let grid = Grid::from(vec![Bins::new(Edges::from(edges_vec))]);
+        
+        // get the histogram
+        let flat = Array::from_iter(c.iter()).insert_axis(Axis(1));
+        let flat2 = flat.mapv(
+            |x| ((*x).to_f64().unwrap() * scale_factor).to_u8().unwrap()
+        );
+        let hist = flat2.histogram(grid);
+        // Straight out of wikipedia: 
+        let counts = hist.counts(); 
+        let total = counts.sum().to_f64().unwrap();
+        let counts = Array::from_iter(counts.iter());
+        // TODO: Could use the following for skimage-esque implementation 
```

Comment:
Remove the commented out code, potentially leave a comment explaining how this could be done